### PR TITLE
PUBDEV-7796: Model explainability interface for H2O models and AutoML objects in R/Python

### PR DESCRIPTION
**Pull request by tomasfryda**
https://0xdata.atlassian.net/browse/PUBDEV-7796

WIP: Near final version, but still making edits

## Explainability API 

Documenting the new methods (for feedback / naming):

Python methods in the `h2o.explain` module:

These take a single model or (including an AutoML object or leaderboard slice) as input:
```
explain        
explain_row        
```

Takes a list of models (including an AutoML object or leaderboard slice) as input:
```
varimp_heatmap          
model_correlation_heatmap        
pd_multi_plot
```

Takes a single model as input:
```
residual_analysis_plot
shap_explain_row_plot
shap_summary_plot
pd_plot
ice_plot
```

R has the same functions names, but with the `h2o.*` prefix.

### CODE REVIEW

File: h2o-r/h2o-DESCRIPTION.template
Comment (tomasfryda) [inline h2o-r/h2o-DESCRIPTION.template:61]: `DT`, `htmltools`, `plotly` are required for html output of explanation.
`IRdisplay`, `repr` is required for html rendering in Jupyter notebook with R kernel. (Those packages are dependencies of R jupyter kernel).

Without adding them to `Suggests`, R check as CRAN would complain.

File: h2o-r/h2o-DESCRIPTION.template
Comment (tomasfryda) [inline h2o-r/h2o-DESCRIPTION.template:108]: I'm wondering if the `zzz.R` shouldn't be the last file in the list.

**Comment by sebhrusen** [comment]
@tomasfryda 2 new files of ~2000 lines that should be reviewed at once? any trick to give us some courage?

File: h2o-py/h2o/mli/_mli.py
Comment (sebhrusen) [inline h2o-py/h2o/mli/_mli.py:140]: yet another dataframe format?
Wondering if it's worth introducing this when 9x% of our users (and 100% of conda users) will have pandas in their Python environment.

Comment (tomasfryda) [reply h2o-py/h2o/mli/_mli.py:140]: Yes, I agree. I would be much happier using pandas instead of my "numpy frame".

**Comment by tomasfryda** [comment]
@sebhrusen I don't know if it helps but significant part of that 2k loc are just comments. Other thing that might help is that the code is not very complex, after all it's just plotting code consisting of getting the data to the right shape and plotting it. 

Python code is slightly more complicated since I couldn't use pandas and scipy (so there is a very simple implementation of density estimation - used just for the jitter in the SHAP summary plot, hierarchical clustering, and the NumpyFrame).

**Comment by ledell** [comment]
@navdeep-G @pramitchoudhary Can you take a look at the descriptions here and suggest/commit any changes/additions? https://github.com/h2oai/h2o-3/blob/e6178699f2dfbed20ea370c54d37e606d9b2f3a1/h2o-py/h2o/mli/_mli.py#L69

File: h2o-r/h2o-DESCRIPTION.template
Comment (ledell) [reply h2o-r/h2o-DESCRIPTION.template:108]: Yeah, let's move it to the end (it used to be there).

File: h2o-r/h2o-DESCRIPTION.template
Comment (ledell) [inline h2o-r/h2o-DESCRIPTION.template:111]: Can we name this file, `explain.R` so it's more obvious what it contains?  I don't think the "MLI" acronym is universally known/used and this code is focused more on explainability vs interpretability.

File: h2o-r/scripts/h2o-r-test-setup.R
Comment (ledell) [inline h2o-r/scripts/h2o-r-test-setup.R:189]: Also change to `"explain.R"`.

File: h2o-r/h2o-package/R/mli.R
Comment (ledell) [inline h2o-r/h2o-package/R/mli.R:1787]: If this is just for plot parameter overrides, then maybe we should use a parameter name like `plot_overrides` or `plot_arguments`?

File: h2o-r/h2o-package/R/mli.R
Comment (ledell) [inline h2o-r/h2o-package/R/mli.R:1774]: @tomasfryda Can you explain this parameter a bit more -- can you fill in the missing pieces:  "If True, explain only the best model from each algorithm class; if False ... "

File: h2o-r/h2o-package/R/mli.R
Comment (ledell) [inline h2o-r/h2o-package/R/mli.R:1769]:
```suggestion
#' @param columns_of_interest A vector of feature names. If specified, plot just these columns.
```

Comment (ledell) [reply h2o-r/h2o-package/R/mli.R:1769]: This is spelled wrong, and will need to be changed elsewhere too.

File: h2o-r/h2o-package/R/mli.R
Comment (ledell) [inline h2o-r/h2o-package/R/mli.R:1773]:
```suggestion
#' @param top_n_features If \code{columns_of_interest} is missing, create plots only with the top n columns (where applicable).  Defaults to 5.
```

File: h2o-r/h2o-package/R/mli.R
Comment (ledell) [inline h2o-r/h2o-package/R/mli.R:1772]:
```suggestion
#' @param exclude_explanations Exclude specified model explanations.
```

File: h2o-r/h2o-package/R/mli.R
Comment (ledell) [inline h2o-r/h2o-package/R/mli.R:1770]:
```suggestion
#' @param include_explanations If specified, return only the specified model explanations.
```

Comment (ledell) [reply h2o-r/h2o-package/R/mli.R:1770]: Will need to be changed elsewhere too.

File: h2o-r/h2o-package/R/mli.R
Comment (ledell) [reply h2o-r/h2o-package/R/mli.R:1772]: Will need to be changed elsewhere too.

File: h2o-r/h2o-DESCRIPTION.template
Comment (ledell) [reply h2o-r/h2o-DESCRIPTION.template:61]: It's ok to add them to Suggests.

File: h2o-r/h2o-DESCRIPTION.template
Comment (ledell) [reply h2o-r/h2o-DESCRIPTION.template:108]: Yes, let's move it there, thanks!

File: h2o-py/h2o/mli/_mli.py
Comment (ledell) [inline h2o-py/h2o/mli/_mli.py:142]: I would refrain from using the term, "MLI", since many people don't know that and we are focused on "explainability" right now.  I wonder if we should rename these modules?  `./explain/explain.py`?

**Comment by ledell** [comment]
If we rename file objects from "mli" to "explain",

and could we have aliases so that this:
```
from h2o.explain import explain, explain_row
```
could be written as :

```
from h2o import explain, explain_row
```
or could that not work?

File: h2o-r/h2o-package/R/mli.R
Comment (tomasfryda) [reply h2o-r/h2o-package/R/mli.R:1787]: will change to `plot_overrides`

File: h2o-r/h2o-package/R/mli.R
Comment (tomasfryda) [reply h2o-r/h2o-package/R/mli.R:1769]: thanks!

File: h2o-r/h2o-package/R/mli.R
Comment (tomasfryda) [reply h2o-r/h2o-package/R/mli.R:1774]: That parameter now applies only to the PDP plot - by default it's showing just the best of family. 
I wonder if I should remove the parameter from the `h2o.explain` - it would still be overridable by the `plot_overrides`.
Or I could use just best of family for the rest of the "multi-model" explanations - model correlation heat map and variable importance heat map - now it uses top 20 models which seems to me to be better so I would probably go for the first option.

What do you think @ledell ?

**Comment by tomasfryda** [comment]
> If we rename file objects from "mli" to "explain",
> 
> could we have aliases so that this:
> 
> ```
> from h2o.explain import explain, explain_row
> ```
> 
> could be written as :
> 
> ```
> from h2o import explain, explain_row
> ```
> 
> or could that not work?

@ledell Renamed it to explain. And `from h2o import explain, explain_row` works as well as `h2o.explain_row(...)` and of course the longer version works too (e.g., `h2o.explain.explain_row(...)`)

File: h2o-r/h2o-package/R/mli.R
Comment (tomasfryda) [reply h2o-r/h2o-package/R/mli.R]: > I wonder if I should remove the parameter from the `h2o.explain` - it would still be overridable by the `plot_overrides`.

Just noticed that in Python version I don't expose this parameter in `h2o.explain` (but the user can still choose to show all models by using the `plot_overrides`)

Comment (tomasfryda) [reply h2o-r/h2o-package/R/mli.R]: I removed this parameter and made it smart as is in python: 
When explaining AutoML, show just best of family to get some general overview, but when explaining a list of models assume that the user actually handpicked the models and actually wants to see them all.

This is now a default behaviour but it is still overridable by using the `plot_overrides`.

File: h2o-py/h2o/__init__.py
Comment (sebhrusen) [inline h2o-py/h2o/__init__.py:90]: are you using a try-except because `explain` module imports optional dependencies?
If so, ar those optional dependencies used by all methods you want to expose? I mean, if possible, I'd rather do imports  for optional dependencies inside the methods.

File: h2o-py/h2o/explain/_explain.py
Comment (sebhrusen) [inline h2o-py/h2o/explain/_explain.py:66]: usually, better implement `__str__` for printing and keep `__repr__` for a parseable representation.

File: h2o-py/h2o/__init__.py
Comment (tomasfryda) [reply h2o-py/h2o/__init__.py:90]: @sebhrusen Well everything that I expose outside the module uses `matplotlib` and `matplotlib` is optional. I can move it inside the methods if you still think it would be better.

File: h2o-py/h2o/explain/_explain.py
Comment (tomasfryda) [reply h2o-py/h2o/explain/_explain.py:66]: Good point! I must have confused myself with the IPython's _repr methods.

File: h2o-py/h2o/__init__.py
Comment (sebhrusen) [reply h2o-py/h2o/__init__.py:90]: if all methods are concerned, then your implementation makes sense.
Just wanted to be sure that we're not preventing users from accessing features that would still work in their env.

File: h2o-py/h2o/__init__.py
Comment (pramitchoudhary) [inline h2o-py/h2o/__init__.py:85]: `partial_dependences` Is it not ok to rename this to `partial_dependence_plots`?

Comment (ledell) [reply h2o-py/h2o/__init__.py:85]: @pramitchoudhary I thought the same thing at first... but it's actually just a single plot with multiple lines and not multiple "plots".  The current name, conceptually, makes sense -- my only concern is that people will get confused about whether it's "dependences" or "dependencies".  Also good to note that we have an existing function called `.partial_plot()` which takes a single model and a column (or columns) to generate a PD plot.

File: h2o-py/h2o/explain/_explain.py
Comment (sebhrusen) [inline h2o-py/h2o/explain/_explain.py:1446]: - `columns_of_interest`: isn't it a bit verbose? why not just `columns`?
- `top_n_features`: serves same purpose as previous param, but we use "columns" in one and "features" in the other. Besides, they're mutually exclusive. Can't they be merged into one param? `features=5` or `features=['col_1', 'col_2']`?
- `plot_overrides`: in matplotlib world, don't we just usually use a `kwargs` to forward  those?
- `plot_overrides`, `exclude_explanations`: careful with mutable defaults (here {}, []), I always need to check that this param is not modified in the impl.
- `include_explanations`, `exclude_explanations`: just an idea, as those explanations are predefined, what aout a single  `explanations` param accepting None (=ALL) or instances of:
```
class Explanations():
     def __init__(self, leaderboard=None, ....):
            # all params default to None
            # if all set to None -> include all
            # otherwise, includes only the ones with value True, or excludes the ones with value False.

#usage
explanations=Explanations(leaderboard=True, model_correlation_heatmap=True)
explanations=Explanations(model_correlation_heatmap=False)
```
Benefits: discoverability+autocompletion for the possible explanations, include+exclude in one param.

Comment (sebhrusen) [reply h2o-py/h2o/explain/_explain.py:1446]: same for `explain_row`

File: h2o-py/h2o/__init__.py
Comment (pramitchoudhary) [reply h2o-py/h2o/__init__.py:85]: Got it. Then why not just call it `partial_dependence_plot` so that its clear one is not getting the values but the actual plot and if one is returning the values then we call it `partial_dependence`?

Also, if the `explain` functions are coming from a different namespace then is it not ok to call it `partial_plot()` for consistency?

File: h2o-py/h2o/__init__.py
Comment (pramitchoudhary) [inline h2o-py/h2o/__init__.py:83]: I don't have a strong opinion at this point but thinking from the top  of my mind,
do we need to name function based on the viz it's generating rather than just a `type` param?
For e.g. `variable_importance(type='heatmap', ...)`. 
That way if a single model is passed then it generates a `bar` plot and if a list of models is passed then it generates a heatmap.

If it's too much effort right now, we make modifications for future releases.

File: h2o-py/h2o/explain/_explain.py
Comment (tomasfryda) [reply h2o-py/h2o/explain/_explain.py:1446]: > * `columns_of_interest`: isn't it a bit verbose? why not just `columns`?
> * `top_n_features`: serves same purpose as previous param, but we use "columns" in one and "features" in the other. Besides, they're mutually exclusive. Can't they be merged into one param? `features=5` or `features=['col_1', 'col_2']`?

we should probably keep the names same in both python and R, right? I can change the behaviour in both. What do you think @ledell?


> * `plot_overrides`: in matplotlib world, don't we just usually use a `kwargs` to forward  those?

this used to be called user overrides, it is used to change parameters such as how many features to plot in one particular type of plot so it's not really the thing that I would use `kwargs` (e.g., line type, color etc)

> * `plot_overrides`, `exclude_explanations`: careful with mutable defaults (here {}, []), I always need to check that this param is not modified in the impl.
 
yep, both are read-only

> * `include_explanations`, `exclude_explanations`: just an idea, as those explanations are predefined, what about a single  `explanations` param accepting None (=ALL) or instances of:
> 
> ```
> class Explanations():
>      def __init__(self, leaderboard=None, ....):
>             # all params default to None
>             # if all set to None -> include all
>             # otherwise, includes only the ones with value True, or excludes the ones with value False.
> 
> #usage
> explanations=Explanations(leaderboard=True, model_correlation_heatmap=True)
> explanations=Explanations(model_correlation_heatmap=False)
> ```
> 
> Benefits: discoverability+autocompletion for the possible explanations, include+exclude in one param.

I like this idea, not sure if I can or should do similar thing in R. Also I'm not sure if we want the method/function arguments differ that much between R and python. Any thoughts @ledell ?

File: h2o-py/h2o/__init__.py
Comment (tomasfryda) [reply h2o-py/h2o/__init__.py:85]: I think using `partial_plot()` could be confusing. (When using jupyter, I often found myself googling the function (just the function name not fully qualified name) instead of waiting for the jupyter's documentation to pop up.)

Comment (pramitchoudhary) [reply h2o-py/h2o/__init__.py:85]: `pdp_plot`?

File: h2o-py/h2o/__init__.py
Comment (tomasfryda) [reply h2o-py/h2o/__init__.py:83]: I like the idea, that way we could also do a box plot for multiple models with the same API etc.

I would probably keep it the way it is right now (not much time before release) and in future make a function/method `variable_importance` that would dispatch depending on the input if that sounds reasonably.

**Comment by pramitchoudhary** [comment]
> @navdeep-G @pramitchoudhary Can you take a look at the descriptions here and suggest/commit any changes/additions?
> 
> https://github.com/h2oai/h2o-3/blob/e6178699f2dfbed20ea370c54d37e606d9b2f3a1/h2o-py/h2o/mli/_mli.py#L69

@ledell the description looks good to me.
Just minor stuff,
1. 
`pdp="Partial dependence plot (PDP)`. Should we add the keyword `one-way` explicitly to the description? to indicate we are talking about One-way PDPs. And use the word global explanation?

2. `ice_row`: will it possible to use the word local explanation?

File: h2o-py/h2o/__init__.py
Comment (pramitchoudhary) [reply h2o-py/h2o/__init__.py:83]: sounds good.

Comment (sebhrusen) [reply h2o-py/h2o/__init__.py:83]: if `variable_importance_heatmap` is the only one supported currently, you can still rename it to `variable_importance` and implement the dispatch later, right? This would prevent you from having to implement deprecations later..

**Review by navdeep-G — APPROVED** [review]
I cannot comment on the API too much (I will leave that to the rest of you), but in terms of functionality, I approve.

File: h2o-py/h2o/explain/_explain.py
Comment (tomasfryda) [reply h2o-py/h2o/explain/_explain.py:1481]: @ledell 
- Changed the `columns_of_interest` and `top_n_features` to one parameter `features` with behaviour as Seb described above in both Python and R.

- In python changed the implementation of explanation to use Seb's idea. Usage:
```
import h2o
from h2o.explain import Explanations, RowExplantations

aml.explain(test, explanations=Explanations(leaderboard=False))
aml.explain_row(test, 1, explanations=RowExplanations(ice=True))
```

In R, it stays the same as before (include_explanations/exclude_explanations). I can easily revert this commit or change somehow R's behaviour.

Comment (ledell) [reply h2o-py/h2o/explain/_explain.py:1481]: Let's change the name `features` to `columns` for consistency with other H2O functions.  We can keep the support for integers (to trigger top N) if that seems better, but I am also not opposed to having an extra arg for the latter.

Comment (ledell) [reply h2o-py/h2o/explain/_explain.py:1481]: @tomasfryda @sebhrusen Regarding changing the format of the explanation args in Python and converting into a single `explanations` parameter -- it sounds like the main benefits are:

- remove a parameter
- autocompletion

It's a handy way of storing that information, but I also see some drawbacks:

- Inconsistency between R and Python (adds complexity / duplicates documentation).
- User has two extra imports
- This is not an established way of specifying params in H2O Python module (it could be in the future, but we don't have anything like that that we know of, so far, so it might seem a bit too unique to some folks...?)
- It requires bout 2x the amount of typing because now you have to put True or False in addition to the name.

I honestly don't love the whole `include_*` `exclude_*` thing, but it's got some "existing conventions" weight to it.

Comment (tomasfryda) [reply h2o-py/h2o/explain/_explain.py:1445]: @ledell seems like you'd prefer the include/exclude thing. I dropped the commit for now but I'm keeping the patch handy in case we would like to use the `Explanation` object after all...

**Comment by ledell** [comment]
Here's some code snippets that can be used to try out the interface, if anyone is interested.  cc  @navdeep-G @pramitchoudhary 


### R 

```
        library(h2o)
        h2o.init()

        # Import a sample binary outcome train/test set into H2O
        train <- h2o.importFile("https://s3.amazonaws.com/erin-data/higgs/higgs_train_10k.csv")
        test <- h2o.importFile("https://s3.amazonaws.com/erin-data/higgs/higgs_test_5k.csv")

        # Identify predictors and response
        y <- "response"
        x <- setdiff(names(train), y)

        # For binary classification, response should be a factor
        train[, y] <- as.factor(train[, y])
        test[, y] <- as.factor(test[, y])

        # Run AutoML
        aml <- h2o.automl(x = x, y = y, 
                          training_frame = train,
                          max_models = 10,
                          seed = 1)

        # Explain leader model & compare with all AutoML models                  
        exa <- h2o.explain(aml, test)
        exa

        # Explain a single H2O model (e.g. leader model from AutoML)
        exm <- h2o.explain(aml@leader, test)
        exm
```

### Python:

```
        import h2o
        from h2o.automl import H2OAutoML
        from h2o.explain import explain, explain_row

        h2o.init()

        # Import a sample binary outcome train/test set into H2O
        train = h2o.import_file("https://s3.amazonaws.com/erin-data/higgs/higgs_train_10k.csv")
        test = h2o.import_file("https://s3.amazonaws.com/erin-data/higgs/higgs_test_5k.csv")

        # Identify predictors and response
        x = train.columns
        y = "response"
        x.remove(y)

        # For binary classification, response should be a factor
        train[y] = train[y].asfactor()
        test[y] = test[y].asfactor()
        
        # Run AutoML
        aml = H2OAutoML(max_models=10, seed=1)
        aml.train(x=x, y=y, training_frame=train)

        # Explain leader model & compare with all AutoML models 
        exa = aml.explain(test)

        # Explain a single H2O model (e.g. leader model from AutoML)
        exm = aml.leader.explain(test)
```

File: scripts/jenkins/jenkinsfiles/Jenkinsfile
Comment (tomasfryda) [inline scripts/jenkins/jenkinsfiles/Jenkinsfile:50]: Revert before merging

File: h2o-py/h2o/__init__.py
Comment (ledell) [reply h2o-py/h2o/__init__.py:85]: We decided to break this into two functions, one that takes a single model, `h2o.pd_plot()` and one that takes multiple models (as a list or AutoML object), `h2o.pd_multi_plot()`.

**Comment by ledell** [comment]
Ready to merge, waiting for tests.

**Review by ledell — APPROVED** [review]

**Pull request merged**
